A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,506 @@
+Note that the new policy will be beneath the `priority_experimental` policy,
+which means that if we pick a host for a session while we are failed over to a
+lower priority and then a higher priority comes back online, we will switch
+hosts at that point and this behavior is different from Envoy.
+
+Let's call the new policy `xds_override_host_experimental`. This policy
+contains subchannel management, endpoint management, and RPC routing logic as
+follows:
+
+* maintain a map - let's call it `address_map`. The key is (IP:port)
+  and the value is the tuple
+  `[subchannel, EDS health-status, list of other equivalent addresses]`.
+  When a host (endpoint) has multiple addresses (e.g. IPv6 vs IPv4 or due to
+  multiple network interfaces) they are said to be equivalent addresses.
+  Whenever we get an updated list of addresses from the resolver, we create an
+  entry in the map for each address, and populate the list of equivalent
+  addresses and EDS health-status. The subchannel is based on subchannel
+  creation/destruction or connection/disconnection from the child policy. The
+  equivalent address list computation on each resolver update is performed as
+  shown in the following pseudo-code:
+
+```
+// Find the transitive closure of all EAGs.
+eags_to_process = [set(eag) for eag in resolver_update]
+completed_eag_sets = []
+all_addresses = set()
+while not eags_to_process.empty():
+  current_eag = eags_to_process.pop_front()
+  for eag in eags_to_process:
+    if eag.intersects(current_eag):  // If they have any elements in common
+      add elements from eag to current_eag
+      remove eag from eags_to_process
+  completed_eag_sets.push_back(current_eag)
+  add elements from current_eag to all_addresses
+// Now completed_eag_sets contains the sets we want.
+// Use that to update the LB policy's map.
+// First, update the equivalent addresses for every address in the update.
+for eag in completed_eag_sets:
+  for address in eag:
+    lb_policy.address_map[address].equivalent_addresses = eag
+// Now remove equivalencies for any address not found in the current update.
+for address, entry in lb_policy.address_map:
+  if address not in all_addresses:
+    entry.equivalent_addresses.clear()
+```
+
+* The `xds_override_host_experimental` policy will always remove addresses in
+  state `DRAINING` when passing an update down to its child policy, regardless
+  of the value of `override_host_status`.
+
+* When the policy gets an update that changes an address's health status to
+  `DRAINING`, if (and only if) the `override_host_status` set includes
+  `DRAINING`, it will avoid closing the connection on that subchannel, even
+  when the child policy tries to do so. Instead, this policy will maintain the
+  connection until either (a) the connection closes due to the server closing
+  it or a network failure, or (b) the policy receives a subsequent EDS update
+  that completely removes the endpoint.
```

Comment:
Seems that the behavior is a little different with envoy's.

> * The `xds_override_host_experimental` policy will always remove addresses in
  state `DRAINING` when passing an update down to its child policy, regardless
  of the value of `override_host_status`.

Envoy never remove a address except that the related ep is removed from the eds list. If the `override_host_status` tells that `DRAINING` host is acceptable, then envoy will mark the  `DRAINING` hosts as selectable by the override logic. It means that these hosts still be selected even the new connections could be created for these hosts. 

> * When the policy gets an update that changes an address's health status to
  `DRAINING`, if (and only if) the `override_host_status` set includes
  `DRAINING`, it will avoid closing the connection on that subchannel, even
  when the child policy tries to do so. Instead, this policy will maintain the
  connection until either (a) the connection closes due to the server closing
  it or a network failure, or (b) the policy receives a subsequent EDS update
  that completely removes the endpoint.

When there is status change to mark a host as unhealhy, envoy always will drains all **exists** connections. But new connections still could be created for it. (It's a little weird to change the health check drain policy according to `override_host_status` for envoy. But we could a new configurate for more specific unhealthy host draining 🤔 ).

For example, if host A is marked be `DRAINING`, all exists connection of A will be drained. And by default, host A will be excluded by the load balancing policy. But host A still could be selected by the override logic for exist sessions and new connections will be created to service exist sessions.

Reply:
The goal here is for the behavior to be the same as in Envoy.  My understanding from talking to @htuch and @yanavlasov is that Envoy's behavior is as follows for `DRAINING` hosts:
- The host is not used for new requests, *except* for the host-override case when `override_host_status` includes `DRAINING`.
- Envoud does *not* proactively close existing connections, although connections may wind up being closed for other reasons (e.g., hitting the MAX_CONCURRENT_STREAMS limit).  (This is different from what you describe above, but I think it's accurate.)
- If the host is chosen by a host-override and there is no existing connection, a new connection will be established.

I think this design provides essentially that same behavior.  However, in order to understand why, I think it's necessary to first understand the fundamental architectural differences between Envoy and gRPC in terms of connection management.

Envoy was designed to have a connection pool for each host, independent of the LB policies, probably because of the need to support HTTP/1.  The LB policy picks from a list of hosts without knowing or caring what connections already exist for those hosts.  Then, once the LB policy has chosen a host, Envoy will check to see if the host already has an existing connection it can reuse, and if not, it creates a new connection on the fly.

In gRPC, however, the LB policies are fully in control of connection management.  The LB policy is given a list of addresses, and it decides when to establish a connection to each address.  Because gRPC is built on HTTP/2, there's generally no more than one connection to a given address at a time.  The LB policy is basically picking which connection the request should be sent on, so it will never pick an address for which there is no existing connection, which allows us to minimize request latency.

In gRPC, the connections themselves are ref-counted.  The LB policy holds a ref to the connection, and a new ref is created for each request that is sent on that connection.  This means that we don't actually need the concept of a draining state.  When the LB policy is given an updated address list that no longer includes one of the addresses, the LB policy will immediately drop its ref to the connection for that address, but the connection will remain open until all currently in-flight requests are completed and drop their refs to the connection.

So, with that background in mind, what this part of the design is describing is essentially how we map Envoy's connectivity semantics to gRPC's, so that we get the same resulting behavior for `DRAINING` endpoints.  Specifically:

- Today, when EDS moves an endpoint to `DRAINING` state, we immediately stop using that endpoint in our LB policies, because we don't want to send any more traffic to them -- but because the gRPC connections are ref-counted, the existing connection to that endpoint won't be closed if there are still requests in flight on that connection.  This design is proposing a change to that behavior, which is that we will retain those endpoints for use in the `override_host` LB policy.  However, the gRPC LB policies that sit beneath the `override_host` policy still will not need to see these endpoints, so the `override_host` policy will remove those addresses before it passes the address list to its child.
- Because the child LB policy will no longer see this address (and will therefore drop its ref to the connection), the `override_host` policy will take over holding a ref to the connection.  Unlike Envoy, it will actually proactively maintain a connection to the address without waiting for a request with a host override to select the address, but the net result is the same: we will still be able to send any requests with a host override to this host.  Note that this behavior applies only if `override_host_status` includes `DRAINING`; if it does not, then we know that we will never send host-override requests to this address, so the `override_host` policy does not need to maintain the connections to them (i.e., it reverts to the existing gRPC behavior for `DRAINING` endpoints).

I hope this all makes sense.  Please let me know if you have any questions about any of this.

Reply:
Thanks for your patience and clear explanation. 👍 It make sense and I think it solved my question.

FWIW,

> Envoy does not proactively close existing connections, although connections may wind up being closed for other reasons (e.g., hitting the MAX_CONCURRENT_STREAMS limit). (This is different from what you describe above, but I think it's accurate.)

There are two different draining policy for Envoy. One used when we prepare to remove a host. One used for unhealthy host.
The second one will close all free connection and mark busy connections as draining (these connections will be closed when they become free). 
IMO, it would be better to make the draining policy configurable. But It's not supported for now.

Reply:
Yeah, I understand that Envoy does do proactive draining for unhealthy hosts.  But as far as I can tell, it does not do that for draining hosts, which is the case that we're dealing with here.

Given the way that gRPC does connection management, I don't think we will need to care about the draining policy, even if Envoy makes it configurable in the future.  But we can look at the details when the time comes.

Reply:
I spent some time looking at the Envoy code and talking with @htuch and @yanavlasov about this, and we could not find any place where it was proactively closing connections for hosts in `DRAINING` state.  It looks like it does that only for active health check failures.

Reply:
you are right. 
I just re-checked the code of eds. Any status change by the eds will not trigger the connection draining.
Seems like an inconsistenciey. l will add this to my Todo list to ensure if it's a bug or design.

Reply:
Right.  Note that if you do change Envoy to proactively drain connections when endpoints enter state `DRAINING`, then you probably want to *not* do that when `override_host_status` includes `DRAINING`, because otherwise you will wind up draining and reestablishing connections frequently for stateful sessions that are still tied to that address.